core/src/Telemetry/TelemetryManager.cs

```diff
@@ -0,0 +1,243 @@
+﻿// ------------------------------------------------------------------------------
+// 
+// Copyright (c) Microsoft Corporation.
+// All rights reserved.
+// 
+// This code is licensed under the MIT License.
+// 
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files(the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+// 
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+// 
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+// 
+// ------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Microsoft.Identity.Core.Telemetry
+{
+    internal class TelemetryManager : ITelemetryManager, ITelemetry
+    {
+        private readonly object _lockObj = new object();
+        private ITelemetryReceiver _telemetryReceiver;
+
+        public TelemetryManager(ITelemetryReceiver telemetryReceiver = null)
+        {
+            _telemetryReceiver = telemetryReceiver;
+        }
+
+        public ITelemetryReceiver TelemetryReceiver
+        {
+            get
+            {
+                lock (_lockObj)
+                {
+                    return _telemetryReceiver;
+                }
+            }
+            set
```

Comment:
I haven't unpacked the code to see where the setter is used, but if we can replace the setter with just building a new object, then TelemetryManager would become immutable and there would be no need for a lock.

Reply:
Our customer developers can set the telemetry receiver at any time, but the TelemetryManager will be constructed and injected into existing ClientApplication instances so we can't just replace the entire TelemetryManager everywhere but we need to update where to route telemetry events to.  

Take a look in ClientApplicationBase line 72, i'm assuming this will eventually become public.  The other option in the public API would be to pass in the TelemetryReceiver as part of ClientApplication construction.  I'd prefer to do this.  But until we can remove the public Telemetry static for setting the receiver whenever someone feels like it (and at a global process level) having this setter (and the appropriate locks) are the only solution I can see.  Once we can break the API a bit for how these get set on a per-PCA basis then we can clean this up.

Reply:
Makes sense, thanks!